src/ast/passes/filter_pass.cpp

```diff
@@ -0,0 +1,146 @@
+#include "ast/passes/filter_pass.h"
+#include "ast/ast.h"
+#include "ast/passes/resolve_imports.h"
+#include "ast/visitor.h"
+
+namespace bpftrace::ast {
+
+namespace {
+
+class FilterPass : public Visitor<FilterPass> {
+public:
+  explicit FilterPass(ASTContext &ast,
+                      Imports &imports,
+                      const FilterInputs &inputs)
+      : ast_(ast), imports_(imports), inputs_(inputs) {};
+
+  using Visitor<FilterPass>::visit;
+  void visit(Probe &probe);
+
+private:
+  ASTContext &ast_;
+  Imports &imports_;
+  const FilterInputs inputs_;
+};
+
+// We only needs filters for probes that may be associated with some relevant
+// userspace process. This is not generally the case for generic hardware &
+// software performance counters, for example.
+bool probe_needs_filter(AttachPoint *ap)
+{
+  ProbeType type = probetype(ap->provider);
+
+  switch (type) {
+    case ProbeType::kprobe:
+    case ProbeType::kretprobe:
+    case ProbeType::fentry:
+    case ProbeType::fexit:
+    case ProbeType::tracepoint:
+    case ProbeType::rawtracepoint:
+      return true;
+    // These probe types support passing the pid during attachment
```

Comment:
Is this also true for cgroup filtering? Shouldn't we use a different `cgroup_needs_filter` function?